exp/runtime/internal/controllers/extensionconfig_controller.go

```diff
@@ -48,18 +56,180 @@ func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, opt
 	if err != nil {
 		return errors.Wrap(err, "failed setting up with a controller manager")
 	}
+	err = mgr.Add(&WarmupRunnable{
+		Client:        r.Client,
+		APIReader:     r.APIReader,
+		RuntimeClient: r.RuntimeClient,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+	return nil
+}
+
+var warmupTimeout = time.Second * 60
+var _ manager.LeaderElectionRunnable = &WarmupRunnable{}
+
+// WarmupRunnable is a controller runtime LeaderElectionRunnable. It warms up the extension registry on controller start.
+type WarmupRunnable struct {
+	Client        client.Client
+	APIReader     client.Reader
+	RuntimeClient runtimeclient.Client
+}
+
+// NeedLeaderElection satisfies the controller runtime LeaderElectionRunnable interface.
+func (r *WarmupRunnable) NeedLeaderElection() bool {
+	return true
+}
+
+// Start attempts to warm up the registry. It will retry for 60 seconds before return an error. An error on Start will
+// cause the CAPI controller manager to fail.
+func (r *WarmupRunnable) Start(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	timeout := time.NewTicker(warmupTimeout)
+	defer timeout.Stop()
+
+	var err error
+	go func() {
+		for {
+			if err = warmupRegistry(ctx, r.Client, r.APIReader, r.RuntimeClient); err != nil {
+				log.Error(err, "ExtensionConfig discovery failed")
+				time.Sleep(2 * time.Second)
+			} else {
+				break
+			}
+		}
+	}()
+
+	// Wait for discovery to finish or timeout.
+	for range timeout.C {
```

Comment:
What do you think of using wait.PollImmediateWithContext instead? I think it's non-trivial to implement it ourselves correctly and wait.PollImmediateWithContext seems to do what we want.

It would look like this:
```go
func (r *WarmupRunnable) Start(ctx context.Context) error {
	log := ctrl.LoggerFrom(ctx)

	ctx, cancel := context.WithTimeout(ctx, warmupTimeout)
	defer cancel()
	
	err := wait.PollImmediateWithContext(ctx, 2*time.Second, warmupTimeout, func(ctx context.Context) (done bool, err error) {
		if err = warmupRegistry(ctx, r.Client, r.APIReader, r.RuntimeClient); err != nil {
			log.Error(err, "ExtensionConfig discovery failed")
			return false, nil
		}
		return true, nil
	})
	
	if err != nil {
		return errors.Wrapf(err, "ExtensionConfig discovery timed out after 60 seconds")
	}
	return nil
}
```

I tested it by adding a http call into the func and simulated a timeout via netcat. The call was canceled when the timeout was reached

nit: if warmupTimeout is a constant, maybe warmupInterval should be one too